File: src/theorie.py
"""
theorie.py - Formules analytiques des files d'attente (pour comparaison).

On compare les résultats simulés aux valeurs théoriques :
  - M/M/1 (file infinie) : formules fermées classiques,
  - M/M/c/K (c serveurs, capacité K) : formule générale de naissance-mort,
    qui couvre aussi M/M/1/K (c=1).
"""

import math


def mm1(lam, mu):
    """M/M/1 file infinie. Retourne (rho, L, Lq, W, Wq) ou None si instable."""
    rho = lam / mu
    if rho >= 1.0:
        return None  # système instable : pas de régime stationnaire
    L = rho / (1 - rho)
    Lq = rho * rho / (1 - rho)
    W = 1.0 / (mu - lam)
    Wq = rho / (mu - lam)
    return {"rho": rho, "L": L, "Lq": Lq, "W": W, "Wq": Wq}


def mmck(lam, mu, c, K):
    """M/M/c/K : c serveurs, capacité totale K (file + service).

    Modèle de naissance-mort fini, toujours stable (états bornés).
    Retourne un dict avec rho (par serveur), L, Lq, W, Wq, P_rejet, lambda_eff.
    """
    a = lam / mu  # charge offerte (intensité de trafic)

    # Probabilités non normalisées des états n = 0..K
    termes = []
    for n in range(K + 1):
        if n < c:
            termes.append(a ** n / math.factorial(n))
        else:
            termes.append(a ** n / (math.factorial(c) * c ** (n - c)))
    somme = sum(termes)
    P = [t / somme for t in termes]  # distribution stationnaire

    P_rejet = P[K]                    # proba que le système soit plein
    lam_eff = lam * (1 - P_rejet)     # taux réellement accepté

    L = sum(n * P[n] for n in range(K + 1))
    # Serveurs occupés en moyenne = lam_eff / mu  ->  Lq = L - serveurs occupés
    Lq = L - lam_eff / mu
    W = L / lam_eff
    Wq = Lq / lam_eff
    rho = a / c                        # taux d'utilisation par serveur

    return {"rho": rho, "L": L, "Lq": Lq, "W": W, "Wq": Wq,
            "P_rejet": P_rejet, "lambda_eff": lam_eff}

File: java/seance3/src/Theorie.java
/**
 * Theorie.java — Formules analytiques des files d'attente (pour comparaison).
 *
 *   - M/M/1 (file infinie) : formules fermées classiques.
 *   - M/M/c/K (c serveurs, capacité K) : modèle de naissance-mort fini, qui
 *     couvre aussi M/M/1/K (c=1).
 */
public class Theorie {

    /** Résultat théorique. L = -1 et W = -1 signalent un système instable. */
    public static class Resultat {
        public double rho, L, Lq, W, Wq, pRejet;
        public boolean stable = true;
    }

    /** M/M/1 file infinie. stable=false si rho >= 1. */
    public static Resultat mm1(double lam, double mu) {
        Resultat r = new Resultat();
        r.rho = lam / mu;
        if (r.rho >= 1.0) {           // instable : pas de régime stationnaire
            r.stable = false;
            r.L = r.Lq = r.W = r.Wq = Double.POSITIVE_INFINITY;
            return r;
        }
        r.L  = r.rho / (1 - r.rho);
        r.Lq = r.rho * r.rho / (1 - r.rho);
        r.W  = 1.0 / (mu - lam);
        r.Wq = r.rho / (mu - lam);
        r.pRejet = 0.0;
        return r;
    }

    /** M/M/c/K : c serveurs, capacité totale K (file + service). Toujours stable. */
    public static Resultat mmck(double lam, double mu, int c, int K) {
        double a = lam / mu;          // charge offerte

        // Probabilités non normalisées des états n = 0..K
        double[] termes = new double[K + 1];
        for (int n = 0; n <= K; n++) {
            if (n < c) termes[n] = Math.pow(a, n) / factorielle(n);
            else       termes[n] = Math.pow(a, n) / (factorielle(c) * Math.pow(c, n - c));
        }
        double somme = 0.0;
        for (double t : termes) somme += t;

        double[] P = new double[K + 1];
        for (int n = 0; n <= K; n++) P[n] = termes[n] / somme;

        Resultat r = new Resultat();
        r.pRejet = P[K];                       // proba système plein
        double lamEff = lam * (1 - r.pRejet);  // taux réellement admis
        double L = 0.0;
        for (int n = 0; n <= K; n++) L += n * P[n];
        r.L  = L;
        r.Lq = L - lamEff / mu;                // serveurs occupés moyens = lamEff/mu
        r.W  = L / lamEff;
        r.Wq = r.Lq / lamEff;
        r.rho = a / c;                         // utilisation par serveur
        return r;
    }

    private static double factorielle(int n) {
        double f = 1.0;
        for (int i = 2; i <= n; i++) f *= i;
        return f;
    }
}
